osu.Game.Tests/Input/RealmKeyBindingStoreTest.cs

```diff
@@ -0,0 +1,117 @@
+                Assert.That(bindings[2].KeyCombination, Is.EqualTo(new KeyCombination(InputKey.F1)));
+
+                Assert.That(bindings[3].Action, Is.EqualTo((int)GlobalAction.MusicNext));
+                Assert.That(bindings[3].KeyCombination, Is.EqualTo(new KeyCombination(InputKey.F5)));
+            });
+        }
+
+        [Test]
+        public void TestDuplicateBindingsAreCleared()
+        {
+            var bindings = new List<RealmKeyBinding>
+            {
+                new RealmKeyBinding(GlobalAction.Back, KeyCombination.FromKey(Key.Escape)),
+                new RealmKeyBinding(GlobalAction.Back, KeyCombination.FromMouseButton(MouseButton.Button1)),
+                new RealmKeyBinding(GlobalAction.MusicPrev, KeyCombination.FromKey(Key.F1)),
+                new RealmKeyBinding(GlobalAction.IncreaseVolume, KeyCombination.FromKey(Key.Escape)),
+                new RealmKeyBinding(GlobalAction.MusicNext, KeyCombination.FromKey(Key.F5)),
+                new RealmKeyBinding(GlobalAction.ExportReplay, KeyCombination.FromKey(Key.F1)),
+                new RealmKeyBinding(GlobalAction.TakeScreenshot, KeyCombination.FromKey(Key.PrintScreen)),
+            };
+
+            int countCleared = RealmKeyBindingStore.ClearDuplicateBindings(bindings);
+
+            Assert.Multiple(() =>
+            {
+                Assert.That(countCleared, Is.EqualTo(4));
+
+                Assert.That(bindings[0].Action, Is.EqualTo((int)GlobalAction.Back));
+                Assert.That(bindings[0].KeyCombination, Is.EqualTo(new KeyCombination(InputKey.None)));
+
+                Assert.That(bindings[1].Action, Is.EqualTo((int)GlobalAction.Back));
+                Assert.That(bindings[1].KeyCombination, Is.EqualTo(new KeyCombination(InputKey.ExtraMouseButton1)));
+
+                Assert.That(bindings[2].Action, Is.EqualTo((int)GlobalAction.MusicPrev));
+                Assert.That(bindings[2].KeyCombination, Is.EqualTo(new KeyCombination(InputKey.None)));
+
+                Assert.That(bindings[3].Action, Is.EqualTo((int)GlobalAction.IncreaseVolume));
+                Assert.That(bindings[3].KeyCombination, Is.EqualTo(new KeyCombination(InputKey.None)));
+
+                Assert.That(bindings[4].Action, Is.EqualTo((int)GlobalAction.MusicNext));
+                Assert.That(bindings[4].KeyCombination, Is.EqualTo(new KeyCombination(InputKey.F5)));
+
+                Assert.That(bindings[5].Action, Is.EqualTo((int)GlobalAction.ExportReplay));
+                Assert.That(bindings[5].KeyCombination, Is.EqualTo(new KeyCombination(InputKey.None)));
+
+                Assert.That(bindings[6].Action, Is.EqualTo((int)GlobalAction.TakeScreenshot));
+                Assert.That(bindings[6].KeyCombination, Is.EqualTo(new KeyCombination(InputKey.PrintScreen)));
+            });
+        }
+
+        [Test]
+        public void TestDuplicateBindingsAllowedIfBoundToSameAction()
```

Comment:
This test could seem like it's exercising an edge case, except for the part where catch actually would get caught on this:

https://github.com/ppy/osu/blob/75011db3a70dbcb177256671c569e613b90e7c19/osu.Game.Rulesets.Catch/CatchRuleset.cs#L56-L57

which would cause default dash bindings to become spontaneously completely unbound.

Reply:
I do wonder if the default for catch should be `InputKey.None` for the second (to create an extra binding but have no default)...

Reply:
Sure, but I believe such a change would not apply retroactively to existing configs (unless I specifically include it in the migration I guess).

Reply:
This does seem to work on a quick test. Shall we change it while we're here?

```diff
diff --git a/osu.Game.Rulesets.Catch/CatchRuleset.cs b/osu.Game.Rulesets.Catch/CatchRuleset.cs
index efd4b46782..63dc01d0d0 100644
--- a/osu.Game.Rulesets.Catch/CatchRuleset.cs
+++ b/osu.Game.Rulesets.Catch/CatchRuleset.cs
@@ -54,7 +54,7 @@ public class CatchRuleset : Ruleset, ILegacyRuleset
             new KeyBinding(InputKey.X, CatchAction.MoveRight),
             new KeyBinding(InputKey.Right, CatchAction.MoveRight),
             new KeyBinding(InputKey.Shift, CatchAction.Dash),
-            new KeyBinding(InputKey.Shift, CatchAction.Dash),
+            new KeyBinding(InputKey.None, CatchAction.Dash),
         };
 
         public override IEnumerable<Mod> ConvertFromLegacyMods(LegacyMods mods)

```

Reply:
As pointed out on twitch chat, we should likely have the defaults match stable, which would make it:

```diff
diff --git a/osu.Game.Rulesets.Catch/CatchRuleset.cs b/osu.Game.Rulesets.Catch/CatchRuleset.cs
index efd4b46782..9ceb78893e 100644
--- a/osu.Game.Rulesets.Catch/CatchRuleset.cs
+++ b/osu.Game.Rulesets.Catch/CatchRuleset.cs
@@ -54,7 +54,7 @@ public class CatchRuleset : Ruleset, ILegacyRuleset
             new KeyBinding(InputKey.X, CatchAction.MoveRight),
             new KeyBinding(InputKey.Right, CatchAction.MoveRight),
             new KeyBinding(InputKey.Shift, CatchAction.Dash),
-            new KeyBinding(InputKey.Shift, CatchAction.Dash),
+            new KeyBinding(InputKey.MouseLeft, CatchAction.Dash),
         };
 
         public override IEnumerable<Mod> ConvertFromLegacyMods(LegacyMods mods)

```

I think it's fine to leave your test and logic in place as it's probably not going to harm anything. Up to you.

Reply:
I've applied the last diff, and also added backwards migration on top (https://github.com/ppy/osu/pull/25152/commits/4cfc95c673f2fe407f04c3884b083a7ab311795e). I initially wasn't gonna bother but not doing so breaks revert-to-default for reasons I'd rather not have to fix so I decided that I might as well migrate over.